File: CreditCardApp/Models/Core/CreditCard.swift
import Foundation
import CoreData

// MARK: - Credit Card Model
struct CreditCard: Codable, Identifiable {
    let id: UUID
    var name: String
    var cardType: CardType
    var rewardCategories: [RewardCategory]
    var quarterlyBonus: QuarterlyBonus?
    var spendingLimits: [SpendingLimit]
    var isActive: Bool
    var createdAt: Date
    var updatedAt: Date
    
    init(id: UUID = UUID(), 
         name: String, 
         cardType: CardType, 
         rewardCategories: [RewardCategory] = [], 
         quarterlyBonus: QuarterlyBonus? = nil, 
         spendingLimits: [SpendingLimit] = [], 
         isActive: Bool = true) {
        self.id = id
        self.name = name
        self.cardType = cardType
        self.rewardCategories = rewardCategories.isEmpty ? cardType.defaultRewards : rewardCategories
        self.quarterlyBonus = quarterlyBonus
        self.spendingLimits = spendingLimits
        self.isActive = isActive
        self.createdAt = Date()
        self.updatedAt = Date()
    }
}

// MARK: - Reward Category
struct RewardCategory: Codable, Identifiable {
    let id: UUID
    var category: SpendingCategory
    var multiplier: Double
    var pointType: PointType
    var isActive: Bool
    
    init(id: UUID = UUID(), 
         category: SpendingCategory, 
         multiplier: Double, 
         pointType: PointType, 
         isActive: Bool = true) {
        self.id = id
        self.category = category
        self.multiplier = multiplier
        self.pointType = pointType
        self.isActive = isActive
    }
}

// MARK: - Quarterly Bonus
struct QuarterlyBonus: Codable {
    var category: SpendingCategory
    var multiplier: Double
    var pointType: PointType
    var limit: Double
    var currentSpending: Double
    var quarter: Int // 1-4
    var year: Int
    
    init(category: SpendingCategory, 
         multiplier: Double, 
         pointType: PointType, 
         limit: Double, 
         currentSpending: Double = 0.0, 
         quarter: Int = Calendar.current.component(.quarter, from: Date()), 
         year: Int = Calendar.current.component(.year, from: Date())) {
        self.category = category
        self.multiplier = multiplier
        self.pointType = pointType
        self.limit = limit
        self.currentSpending = currentSpending
        self.quarter = quarter
        self.year = year
    }
}

// MARK: - Spending Limit
struct SpendingLimit: Codable, Identifiable {
    let id: UUID
    var category: SpendingCategory
    var limit: Double
    var currentSpending: Double
    var resetDate: Date
    var resetType: ResetType
    
    init(id: UUID = UUID(), 
         category: SpendingCategory, 
         limit: Double, 
         currentSpending: Double = 0.0, 
         resetDate: Date = Date(), 
         resetType: ResetType = .annually) {
        self.id = id
        self.category = category
        self.limit = limit
        self.currentSpending = currentSpending
        self.resetDate = resetDate
        self.resetType = resetType
    }
    
    var usagePercentage: Double {
        guard limit > 0 else { return 0.0 }
        return currentSpending / limit
    }
    
    var remainingAmount: Double {
        return max(0, limit - currentSpending)
    }
    
    var isLimitReached: Bool {
        return currentSpending >= limit
    }
    
    var isWarningThreshold: Bool {
        return usagePercentage >= 0.85
    }
}

// MARK: - User Preferences
struct UserPreferences: Codable {
    var preferredPointSystem: PointType
    var alertThreshold: Double // Percentage (0.85 = 85%)
    var language: Language
    var notificationsEnabled: Bool
    var autoUpdateSpending: Bool
    
    init(preferredPointSystem: PointType = .membershipRewards,
         alertThreshold: Double = 0.85,
         language: Language = .english,
         notificationsEnabled: Bool = true,
         autoUpdateSpending: Bool = false) {
        self.preferredPointSystem = preferredPointSystem
        self.alertThreshold = alertThreshold
        self.language = language
        self.notificationsEnabled = notificationsEnabled
        self.autoUpdateSpending = autoUpdateSpending
    }
}

// MARK: - Chat Message
struct ChatMessage: Codable, Identifiable {
    let id: UUID
    var content: String
    var sender: MessageSender
    var timestamp: Date
    var cardRecommendations: [CardRecommendation]?
    var spendingUpdate: SpendingUpdate?
    
    init(id: UUID = UUID(),
         content: String,
         sender: MessageSender,
         timestamp: Date = Date(),
         cardRecommendations: [CardRecommendation]? = nil,
         spendingUpdate: SpendingUpdate? = nil) {
        self.id = id
        self.content = content
        self.sender = sender
        self.timestamp = timestamp
        self.cardRecommendations = cardRecommendations
        self.spendingUpdate = spendingUpdate
    }
}

enum MessageSender: String, Codable {
    case user = "User"
    case assistant = "Assistant"
}

// MARK: - Card Recommendation
struct CardRecommendation: Codable, Identifiable {
    let id: UUID
    var cardId: UUID
    var cardName: String
    var category: SpendingCategory
    var multiplier: Double
    var pointType: PointType
    var reasoning: String
    var currentSpending: Double
    var limit: Double
    var isLimitReached: Bool
    var rank: Int // 1 for primary, 2 for secondary
    
    init(id: UUID = UUID(),
         cardId: UUID,
         cardName: String,
         category: SpendingCategory,
         multiplier: Double,
         pointType: PointType,
         reasoning: String,
         currentSpending: Double = 0.0,
         limit: Double = 0.0,
         isLimitReached: Bool = false,
         rank: Int = 1) {
        self.id = id
        self.cardId = cardId
        self.cardName = cardName
        self.category = category
        self.multiplier = multiplier
        self.pointType = pointType
        self.reasoning = reasoning
        self.currentSpending = currentSpending
        self.limit = limit
        self.isLimitReached = isLimitReached
        self.rank = rank
    }
}

// MARK: - Spending Update
struct SpendingUpdate: Codable {
    var cardId: UUID
    var category: SpendingCategory
    var amount: Double
    var previousAmount: Double
    
    init(cardId: UUID, category: SpendingCategory, amount: Double, previousAmount: Double) {
        self.cardId = cardId
        self.category = category
        self.amount = amount
        self.previousAmount = previousAmount
    }
}

// MARK: - Merchant
struct Merchant: Codable, Identifiable {
    let id: UUID
    var name: String
    var category: SpendingCategory
    var aliases: [String] // Alternative names
    var isActive: Bool
    
    init(id: UUID = UUID(), name: String, category: SpendingCategory, aliases: [String] = [], isActive: Bool = true) {
        self.id = id
        self.name = name
        self.category = category
        self.aliases = aliases
        self.isActive = isActive
    }
}

// MARK: - Pre-defined Merchant Mappings
let merchantMappings: [String: SpendingCategory] = [
    "costco": .costco,
    "whole foods": .wholeFoods,
    "wholefoods": .wholeFoods,
    "amazon": .amazon,
    "target": .target,
    "walmart": .walmart,
    "starbucks": .coffee,
    "mcdonalds": .fastFood,
    "mcdonald's": .fastFood,
    "uber": .transit,
    "lyft": .transit,
    "netflix": .streaming,
    "spotify": .streaming,
    "cvs": .drugstores,
    "walgreens": .drugstores,
    "shell": .gas,
    "exxon": .gas,
    "chevron": .gas,
    "bp": .gas,
    "mobil": .gas
] 

File: credit_card_app_flutter/lib/models/credit_card.dart
import 'package:uuid/uuid.dart';
import 'spending_category.dart';
import 'point_type.dart';
import 'card_type.dart';

const _uuid = Uuid();

class CreditCard {
  final String id;
  String name;
  CardType cardType;
  List<RewardCategory> rewardCategories;
  QuarterlyBonus? quarterlyBonus;
  List<SpendingLimit> spendingLimits;
  bool isActive;
  final DateTime createdAt;
  DateTime updatedAt;

  CreditCard({
    String? id,
    required this.name,
    required this.cardType,
    List<RewardCategory>? rewardCategories,
    this.quarterlyBonus,
    List<SpendingLimit>? spendingLimits,
    this.isActive = true,
    DateTime? createdAt,
    DateTime? updatedAt,
  })  : id = id ?? _uuid.v4(),
        rewardCategories = (rewardCategories != null && rewardCategories.isNotEmpty)
            ? rewardCategories
            : cardType.defaultRewards,
        spendingLimits = spendingLimits ?? [],
        createdAt = createdAt ?? DateTime.now(),
        updatedAt = updatedAt ?? DateTime.now();

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'cardType': cardType.name,
        'rewardCategories': rewardCategories.map((r) => r.toJson()).toList(),
        'quarterlyBonus': quarterlyBonus?.toJson(),
        'spendingLimits': spendingLimits.map((s) => s.toJson()).toList(),
        'isActive': isActive,
        'createdAt': createdAt.toIso8601String(),
        'updatedAt': updatedAt.toIso8601String(),
      };

  factory CreditCard.fromJson(Map<String, dynamic> json) {
    final ct = CardType.values.firstWhere(
      (e) => e.name == json['cardType'],
      orElse: () => CardType.custom,
    );
    return CreditCard(
      id: json['id'] as String,
      name: json['name'] as String,
      cardType: ct,
      rewardCategories: (json['rewardCategories'] as List?)
              ?.map((r) => RewardCategory.fromJson(r as Map<String, dynamic>))
              .toList() ??
          ct.defaultRewards,
      quarterlyBonus: json['quarterlyBonus'] != null
          ? QuarterlyBonus.fromJson(json['quarterlyBonus'] as Map<String, dynamic>)
          : null,
      spendingLimits: (json['spendingLimits'] as List?)
              ?.map((s) => SpendingLimit.fromJson(s as Map<String, dynamic>))
              .toList() ??
          [],
      isActive: json['isActive'] as bool? ?? true,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  CreditCard copyWith({
    String? name,
    CardType? cardType,
    List<RewardCategory>? rewardCategories,
    QuarterlyBonus? quarterlyBonus,
    List<SpendingLimit>? spendingLimits,
    bool? isActive,
  }) {
    return CreditCard(
      id: id,
      name: name ?? this.name,
      cardType: cardType ?? this.cardType,
      rewardCategories: rewardCategories ?? this.rewardCategories,
      quarterlyBonus: quarterlyBonus ?? this.quarterlyBonus,
      spendingLimits: spendingLimits ?? this.spendingLimits,
      isActive: isActive ?? this.isActive,
      createdAt: createdAt,
      updatedAt: DateTime.now(),
    );
  }
}

class RewardCategory {
  final String id;
  SpendingCategory category;
  double multiplier;
  PointType pointType;
  bool isActive;

  RewardCategory({
    String? id,
    required this.category,
    required this.multiplier,
    required this.pointType,
    this.isActive = true,
  }) : id = id ?? _uuid.v4();

  Map<String, dynamic> toJson() => {
        'id': id,
        'category': category.name,
        'multiplier': multiplier,
        'pointType': pointType.name,
        'isActive': isActive,
      };

  factory RewardCategory.fromJson(Map<String, dynamic> json) {
    return RewardCategory(
      id: json['id'] as String,
      category: SpendingCategory.values.firstWhere(
        (e) => e.name == json['category'],
        orElse: () => SpendingCategory.general,
      ),
      multiplier: (json['multiplier'] as num).toDouble(),
      pointType: PointType.values.firstWhere(
        (e) => e.name == json['pointType'],
        orElse: () => PointType.cashBack,
      ),
      isActive: json['isActive'] as bool? ?? true,
    );
  }
}

class QuarterlyBonus {
  SpendingCategory category;
  double multiplier;
  PointType pointType;
  double limit;
  double currentSpending;
  int quarter;
  int year;

  QuarterlyBonus({
    required this.category,
    required this.multiplier,
    required this.pointType,
    required this.limit,
    this.currentSpending = 0.0,
    int? quarter,
    int? year,
  })  : quarter = quarter ?? _currentQuarter,
        year = year ?? DateTime.now().year;

  static int get _currentQuarter => ((DateTime.now().month - 1) ~/ 3) + 1;

  double get usagePercentage {
    if (limit <= 0) return 0.0;
    return (currentSpending / limit).clamp(0.0, 1.0);
  }

  double get remainingAmount => (limit - currentSpending).clamp(0.0, double.infinity);
  bool get isLimitReached => currentSpending >= limit;

  Map<String, dynamic> toJson() => {
        'category': category.name,
        'multiplier': multiplier,
        'pointType': pointType.name,
        'limit': limit,
        'currentSpending': currentSpending,
        'quarter': quarter,
        'year': year,
      };

  factory QuarterlyBonus.fromJson(Map<String, dynamic> json) {
    return QuarterlyBonus(
      category: SpendingCategory.values.firstWhere(
        (e) => e.name == json['category'],
        orElse: () => SpendingCategory.general,
      ),
      multiplier: (json['multiplier'] as num).toDouble(),
      pointType: PointType.values.firstWhere(
        (e) => e.name == json['pointType'],
        orElse: () => PointType.cashBack,
      ),
      limit: (json['limit'] as num).toDouble(),
      currentSpending: (json['currentSpending'] as num?)?.toDouble() ?? 0.0,
      quarter: json['quarter'] as int,
      year: json['year'] as int,
    );
  }
}

class SpendingLimit {
  final String id;
  SpendingCategory category;
  double limit;
  double currentSpending;
  DateTime resetDate;
  ResetType resetType;

  SpendingLimit({
    String? id,
    required this.category,
    required this.limit,
    this.currentSpending = 0.0,
    DateTime? resetDate,
    this.resetType = ResetType.annually,
  })  : id = id ?? _uuid.v4(),
        resetDate = resetDate ?? DateTime.now();

  double get usagePercentage {
    if (limit <= 0 || currentSpending < 0) return 0.0;
    final pct = currentSpending / limit;
    return pct.isNaN || pct.isInfinite ? 0.0 : pct;
  }

  double get remainingAmount => (limit - currentSpending).clamp(0.0, double.infinity);
  bool get isLimitReached => currentSpending >= limit;
  bool get isWarningThreshold => usagePercentage >= 0.85;

  Map<String, dynamic> toJson() => {
        'id': id,
        'category': category.name,
        'limit': limit,
        'currentSpending': currentSpending,
        'resetDate': resetDate.toIso8601String(),
        'resetType': resetType.name,
      };

  factory SpendingLimit.fromJson(Map<String, dynamic> json) {
    return SpendingLimit(
      id: json['id'] as String,
      category: SpendingCategory.values.firstWhere(
        (e) => e.name == json['category'],
        orElse: () => SpendingCategory.general,
      ),
      limit: (json['limit'] as num).toDouble(),
      currentSpending: (json['currentSpending'] as num?)?.toDouble() ?? 0.0,
      resetDate: DateTime.parse(json['resetDate'] as String),
      resetType: ResetType.values.firstWhere(
        (e) => e.name == json['resetType'],
        orElse: () => ResetType.annually,
      ),
    );
  }
}

enum ResetType {
  monthly('Monthly'),
  quarterly('Quarterly'),
  annually('Annually'),
  never('Never');

  const ResetType(this.displayName);
  final String displayName;
}

class CardRecommendation {
  final String id;
  final String cardId;
  final String cardName;
  final CardType? cardType;
  final SpendingCategory category;
  final double multiplier;
  final PointType pointType;
  final String reasoning;
  final double currentSpending;
  final double limit;
  final bool isLimitReached;
  final int rank;

  CardRecommendation({
    String? id,
    required this.cardId,
    required this.cardName,
    this.cardType,
    required this.category,
    required this.multiplier,
    required this.pointType,
    required this.reasoning,
    this.currentSpending = 0.0,
    this.limit = 0.0,
    this.isLimitReached = false,
    this.rank = 1,
  }) : id = id ?? _uuid.v4();
}
